latex multilingual and contentsname setup to after getting user setup

jfbu opened the pull request:
This may fix both of #3123 and #3069. I simply moved <strike>relevant</strike> code to after

    self.elements.update(builder.config.latex_elements)

(edit) *to fix #3069 and I added a check for user config of babel for #3123* (we must obey ``'babel'`` key for backwards compatibility as it is documented)

Not confident that no after-effect.

jfbu committed:
latex multilingual and contentsname setup to after getting user setup

jfbu commented:
I have amended the commit. Initial version was wrong, because it would overwrite in some cases other user-set keys.

jfbu commented on sphinx/writers/latex.py:
```diff
@@ -462,12 +465,17 @@ def declare_package(packagename, options=None):
         self.elements.update(builder.config.latex_elements)
+
+        if self.elements['babel'] != '3.1415':
+            self.elements['multilingual'] = self.elements['babel']
```
people who in the past use `'babel'` key for `\\usepackage{polyglossia}` with xelatex/lualatex will be ok. But if they set the `'babel'` to empty and used `'preamble'` key for polyglossia, it is no good.

this is complicated: one option would be to say in the documentation that people with xelatex/lualatex want to control exactly the way they use babel or polyglossia **must** use `'babel'` key for that.

jfbu replied:
well, no those who set the 'babel' to empty and used 'preamble' key for polyglossia, are **also ok**.

jfbu replied:
I think I am a bit tired here and can not think straight anymore... I think this trick is fine. Use of `'babel'` by user must overrule latex writer default choices, independently of `latex_engine`.

jfbu committed:
reset ``'babel'`` to empty if it was for ``babel_renewcommand`` ok

tk0miya commented:
Note: I will review this after 1.5b1 release (because this is a bugfix :-)

jfbu commented:
@tk0miya delay will serve as a test of beta-testers ;-)

tk0miya commented on sphinx/writers/latex.py:
```diff
@@ -462,12 +468,20 @@ def declare_package(packagename, options=None):
-                self.babel_renewcommand('\\contentsname', document.settings.contentsname)
         # allow the user to override them all
         self.check_latex_elements()
         self.elements.update(builder.config.latex_elements)
```
Sorry for late response. I reviewed this now.
How about move this line after initialization of `self.elements`?
Then the trick like `3.1415` is not needed.

There are only two keys which are overrided by writer before this line; `fncychap` and `fontpkg`.
So it will go better to prevent overriding.

```
        if builder.config.language and 'fncychap' not in builder.config.latex_elements:
            # use Sonny style if any language specified
            self.elements['fncychap'] = '\\usepackage[Sonny]{fncychap}'
```

```
                # Times fonts don't work with Cyrillic languages
                if self.babel.uses_cyrillic() and 'fontpkg' not in builder.config.latex_elements:
                    self.elements['fontpkg'] = ''
```

Until `multilingual` introduced, it was reasonable to apply user settings after setting babel.
Because setting babel does not depends on user settings.
But, from 1.5, setting babel depends on user settings. So I think moving the process to top is simple and understandable.

jfbu committed:
refactor multilingual (load user settings on top)

jfbu committed:
polyglossia honors ``\addto\captions<language>`` like babel

jfbu commented:
@tk0miya thanks for advice, I tried to implement, can you review please
- I am not sure about set-up for Japanese documents (should user settings been forcefully overwritten always like currently for `fncychap` ?)
- we must honor `'babel'` user setting as the key is public; and for example for French language, it is possible that babel offers better support than polyglossia, hence I added to xelatex config nil setting of `'babel'` key to detect if user has modified it,
- under xelatex+polyglossia, we need `babel_renewcommand` to also use `\addto\captions...` (else e.g. `:caption:` option to `toctree` directive is not obeyed)
- should I rebase this on 1.5-release ?

jfbu committed:
make shorthandoff extra safe also for turkish

(perhaps xelatex+babel+turkish may not use active ``=`` either now
or in future)

tk0miya commented:
Thank you for update. Looks great!

> I am not sure about set-up for Japanese documents (should user settings been forcefully overwritten always like currently for fncychap ?)

I'm not sure too. But this behavior is introduced at 7b15082a (at 2012).
So please leave it as it is

> should I rebase this on 1.5-release ?

I just rebased using web interface :-)
